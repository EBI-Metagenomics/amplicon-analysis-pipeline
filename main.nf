include { validateParameters; paramsHelp; paramsSummaryLog } from 'plugin/nf-validation'

// Print help message, supply typical command line usage for the pipeline
if (params.help) {
   log.info paramsHelp("nextflow run main.nf --input samplesheet.csv --outdir ./outdir")
   exit 0
}
// Validate input parameters
validateParameters()

// Print summary of supplied parameters
log.info paramsSummaryLog(workflow)

include { AMPLICON_PIPELINE } from './workflows/pipeline.nf'

workflow {
    AMPLICON_PIPELINE()
}