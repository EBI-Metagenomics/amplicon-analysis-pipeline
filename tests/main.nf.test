nextflow_pipeline {

    name "Test Workflow AMPLICON_PIPELINE"
    script "main.nf"

    test("Samplesheet amplicon-pipeline") {

        tag "samplesheet"

        when {
            params {
                input  = "${projectDir}/tests/samplesheets/test_samplesheet.csv"
                outdir = "$outputDir"
                non_insdc = true
            }
        }

        then {

            def stable_name = getAllFilesFromDir(params.outdir, relative: true, includeDir: true, ignore: ['pipeline_info/*.{html,yml}'])
            def stable_path = getAllFilesFromDir(params.outdir, ignoreFile: 'tests/.nftignore')
            def compressed_fastq = getAllFilesFromDir(params.outdir, include: ['ERR4334351/qc/*.merged.fastq.gz'])

            // the order of lines change for these files so getting the name + filesize instead
            def mseqs = getAllFilesFromDir(params.outdir, include: ['**/taxonomy-summary/**/*.mseq'])
            def mseq_strings = mseqs.collect { mseq ->
                "${mseq.name} - filesize: ${mseq.size()}"
            }
            // the order of lines change for these files so getting the name + filesize instead
            def asv_taxa = getAllFilesFromDir(params.outdir, include: ['**/asv/*asv_tax.tsv'])
            def asv_taxa_strings = asv_taxa.collect { asv_tax ->
                "${asv_tax.name} - filesize: ${asv_tax.size()}"
            }

            with(workflow) {
                assertAll(
                    { assert success }, 
                    // Test expected task success counts at potential branch points of the pipeline
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:READS_QC_MERGE:SEQFU_CHECK") } == 4 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:READS_QC_MERGE:FASTQSUFFIXHEADERCHECK") } == 3 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:READS_QC_MERGE:PIMENTO_GENERATEBCV") } == 2 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:READS_QC_MERGE:FASTP") } == 2 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:READS_QC_MERGE:SEQTK_SEQ") } == 2 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:RRNA_EXTRACTION:INFERNAL_CMSEARCH") } == 1 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:AMP_REGION_INFERENCE:EXTRACT_VAR_REGIONS") } == 1 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:CONCAT_PRIMER_CUTADAPT:CUTADAPT") } == 1 },
                    { assert trace.succeeded().count{ task -> task.name.contains("AMPLICON_PIPELINE:DADA2_SWF:DADA2") } == 1 },
                    // Test output of certain important files
                    { assert snapshot(
                        stable_name,
                        stable_path,
                        compressed_fastq,
                        mseq_strings,
                        asv_taxa_strings
                    ).match() }
                )
            }
        }
    }
}