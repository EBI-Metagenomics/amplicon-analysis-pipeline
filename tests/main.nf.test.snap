{
    "Samplesheet amplicon-pipeline": {
        "content": [
            [
                "ERR4334351",
                "ERR4334351/amplified-region-inference",
                "ERR4334351/amplified-region-inference/ERR4334351.16S.V3-V4.txt",
                "ERR4334351/amplified-region-inference/ERR4334351.tsv",
                "ERR4334351/asv",
                "ERR4334351/asv/16S-V3-V4",
                "ERR4334351/asv/16S-V3-V4/ERR4334351_16S-V3-V4_asv_read_counts.tsv",
                "ERR4334351/asv/ERR4334351_DADA2-PR2_asv_tax.tsv",
                "ERR4334351/asv/ERR4334351_DADA2-SILVA_asv_tax.tsv",
                "ERR4334351/asv/ERR4334351_asv_seqs.fasta",
                "ERR4334351/asv/ERR4334351_dada2_stats.tsv",
                "ERR4334351/primer-identification",
                "ERR4334351/primer-identification/ERR4334351.cutadapt.json",
                "ERR4334351/primer-identification/ERR4334351_primer_validation.tsv",
                "ERR4334351/primer-identification/fwd_primers.fasta",
                "ERR4334351/primer-identification/rev_primers.fasta",
                "ERR4334351/qc",
                "ERR4334351/qc/ERR4334351.fastp.json",
                "ERR4334351/qc/ERR4334351.merged.fastq.gz",
                "ERR4334351/qc/ERR4334351_multiqc_report.html",
                "ERR4334351/qc/ERR4334351_seqfu.tsv",
                "ERR4334351/qc/ERR4334351_suffix_header_err.json",
                "ERR4334351/sequence-categorisation",
                "ERR4334351/sequence-categorisation/ERR4334351.tblout.deoverlapped",
                "ERR4334351/sequence-categorisation/ERR4334351_SSU.fasta",
                "ERR4334351/sequence-categorisation/ERR4334351_SSU_rRNA_archaea.RF01959.fasta",
                "ERR4334351/sequence-categorisation/ERR4334351_SSU_rRNA_bacteria.RF00177.fasta",
                "ERR4334351/taxonomy-summary",
                "ERR4334351/taxonomy-summary/DADA2-PR2",
                "ERR4334351/taxonomy-summary/DADA2-PR2/ERR4334351_16S-V3-V4.html",
                "ERR4334351/taxonomy-summary/DADA2-PR2/ERR4334351_16S-V3-V4_DADA2-PR2_asv_krona_counts.txt",
                "ERR4334351/taxonomy-summary/DADA2-PR2/ERR4334351_DADA2-PR2.mseq",
                "ERR4334351/taxonomy-summary/DADA2-SILVA",
                "ERR4334351/taxonomy-summary/DADA2-SILVA/ERR4334351_16S-V3-V4.html",
                "ERR4334351/taxonomy-summary/DADA2-SILVA/ERR4334351_16S-V3-V4_DADA2-SILVA_asv_krona_counts.txt",
                "ERR4334351/taxonomy-summary/DADA2-SILVA/ERR4334351_DADA2-SILVA.mseq",
                "ERR4334351/taxonomy-summary/PR2",
                "ERR4334351/taxonomy-summary/PR2/ERR4334351.html",
                "ERR4334351/taxonomy-summary/PR2/ERR4334351_PR2.mseq",
                "ERR4334351/taxonomy-summary/PR2/ERR4334351_PR2.tsv",
                "ERR4334351/taxonomy-summary/PR2/ERR4334351_PR2.txt",
                "ERR4334351/taxonomy-summary/SILVA-SSU",
                "ERR4334351/taxonomy-summary/SILVA-SSU/ERR4334351.html",
                "ERR4334351/taxonomy-summary/SILVA-SSU/ERR4334351_SILVA-SSU.mseq",
                "ERR4334351/taxonomy-summary/SILVA-SSU/ERR4334351_SILVA-SSU.tsv",
                "ERR4334351/taxonomy-summary/SILVA-SSU/ERR4334351_SILVA-SSU.txt",
                "ERRNOREADS",
                "ERRNOREADS/qc",
                "ERRNOREADS/qc/ERRNOREADS.fastp.json",
                "ERRNOREADS/qc/ERRNOREADS.merged.fastq.gz",
                "ERRNOREADS/qc/ERRNOREADS_multiqc_report.html",
                "ERRNOREADS/qc/ERRNOREADS_seqfu.tsv",
                "ERRNOREADS/qc/ERRNOREADS_suffix_header_err.json",
                "ERRSEQFUFAIL",
                "ERRSEQFUFAIL/qc",
                "ERRSEQFUFAIL/qc/ERRSEQFUFAIL_seqfu.tsv",
                "ERRSFXHDFAIL",
                "ERRSFXHDFAIL/qc",
                "ERRSFXHDFAIL/qc/ERRSFXHDFAIL_seqfu.tsv",
                "ERRSFXHDFAIL/qc/ERRSFXHDFAIL_suffix_header_err.json",
                "bco.json",
                "pipeline_info",
                "primer_validation_summary.json",
                "qc_failed_runs.csv",
                "qc_passed_runs.csv",
                "study_multiqc_report.html"
            ],
            [
                "ERR4334351.16S.V3-V4.txt:md5,184deb4ff99714b4c71ebc9f2e75eac7",
                "ERR4334351.tsv:md5,3ddd8162ce28f63c89e770f1d15065f4",
                "ERR4334351_16S-V3-V4_asv_read_counts.tsv:md5,bbe6d470d8328746fdead314d727672f",
                "ERR4334351_asv_seqs.fasta:md5,b7aae7dd872f192639b33cf8d25cf1ec",
                "ERR4334351_dada2_stats.tsv:md5,f7921090cba5fe24fff77d3a529dccaf",
                "ERR4334351.cutadapt.json:md5,bd2db828792e12a9296e6c5572982717",
                "ERR4334351_primer_validation.tsv:md5,438b00766b1f24f02de89c99358f8972",
                "fwd_primers.fasta:md5,d3e8e54f8760c09532deaf1f699bd2e0",
                "rev_primers.fasta:md5,85be59f336f4dd4b0419bfcfd62697fa",
                "ERR4334351.fastp.json:md5,4e45f8f12ae5f53f2a4fe59c030cf2e6",
                "ERR4334351_seqfu.tsv:md5,76e1391d4e77b52a44cbaffa6c77975a",
                "ERR4334351_suffix_header_err.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                "ERR4334351.tblout.deoverlapped:md5,b64fff6e87a05f29cec264f7ad52c0d7",
                "ERR4334351_SSU.fasta:md5,3f443c5f8307000620183422284ba015",
                "ERR4334351_SSU_rRNA_archaea.RF01959.fasta:md5,b9a56b94c9d5719828be6da164271083",
                "ERR4334351_SSU_rRNA_bacteria.RF00177.fasta:md5,20a2d2a0afbfe3d5067ad90fc2025b10",
                "ERR4334351_16S-V3-V4_DADA2-PR2_asv_krona_counts.txt:md5,64ef6ae556c1c4565fce9315fe2b05a4",
                "ERR4334351_16S-V3-V4_DADA2-SILVA_asv_krona_counts.txt:md5,791581bb2ca4a3a4f7dbd9e430770071",
                "ERR4334351_PR2.tsv:md5,956e495e06f8fa19fa7073422c00dcdc",
                "ERR4334351_PR2.txt:md5,f2bad5ba0bbe35506039a089383e3e46",
                "ERR4334351_SILVA-SSU.tsv:md5,d3a2880260f198b96e0570e500b678a7",
                "ERR4334351_SILVA-SSU.txt:md5,fa97f9a0cbbfdcc8737e65667e56c6c2",
                "ERRNOREADS.fastp.json:md5,a93f1284ac27748c45d6eaac236785d4",
                "ERRNOREADS_seqfu.tsv:md5,23318103da5360a407d8b5d5ec8784f8",
                "ERRNOREADS_suffix_header_err.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                "ERRSEQFUFAIL_seqfu.tsv:md5,ecab44c2ab1eeeebbc1ae2d9641d60de",
                "ERRSFXHDFAIL_seqfu.tsv:md5,abab2633f525f5cf46e7a94e909dad3d",
                "ERRSFXHDFAIL_suffix_header_err.json:md5,54329c5dd0f5eb4283c60b5cf1b6f3db",
                "primer_validation_summary.json:md5,e3ded4100c30404979344d8987153e39",
                "qc_failed_runs.csv:md5,90b6b3047067e15092481662ba6ba1e7",
                "qc_passed_runs.csv:md5,448f227d11961ca0ad936244c3604547"
            ],
            [
                "ERR4334351.merged.fastq.gz:md5,5c0604ac51f6530b68270a10029a50db"
            ],
            [
                "ERR4334351_DADA2-PR2.mseq - filesize: 11019",
                "ERR4334351_DADA2-SILVA.mseq - filesize: 10782",
                "ERR4334351_PR2.mseq - filesize: 773551",
                "ERR4334351_SILVA-SSU.mseq - filesize: 690708"
            ],
            [
                "ERR4334351_DADA2-PR2_asv_tax.tsv - filesize: 6111",
                "ERR4334351_DADA2-SILVA_asv_tax.tsv - filesize: 4997"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T18:34:39.328859"
    }
}